SIPS/sip-5.md

```diff
@@ -0,0 +1,176 @@
+
+This SIP specifies permission named `snap_signJwt`. This permission grants a snap the ability to create signed JWTs by introducing an additional signing method.
+
+This specification is specified as follows in `snap.manifest.json` files:
+```typescript
+{    
+    "initialPermissions": {
+        "snap_signJwt": {},
+    }
+}
+```
+
+### Common Types
+
+The below common types are used throughout the specification.
+
+```typescript
+interface JWTHeader {
+  typ: string;
+  alg: string;
+  [params: string]: any;
+}
+
+interface JWTPayload {
+  iss?: string;
+  sub?: string;
+  aud?: string | string[];
+  jti?: string;
+  nbf?: number;
+  exp?: number;
+  iat?: number;
+  [claims: string]: any;
+}
+
+type HashingFunction = (data: string | Uint32Array) => string | Uint32Array;
+```
+
+### JWT Signing Implementation
+
+This implementation is modeled after implementations from libraries [did-jwt](https://github.com/decentralized-identity/did-jwt) and [jose](https://github.com/panva/jose).
+
+```typescript
+import * as u8a from "uint8arrays";
+
+function hexToBytes(s: string): Uint8Array {
+  const input = s.startsWith("0x") ? s.substring(2) : s;
+  return u8a.fromString(input.toLowerCase(), "base16");
+}
+
+function bytesToBase64url(b: Uint8Array): string {
+  return u8a.toString(b, "base64url");
+}
+
+function encodeSection(data: any): string {
+  return encodeBase64url(JSON.stringify(data));
+}
+
+export function encodeBase64url(s: string): string {
+  return bytesToBase64url(u8a.fromString(s));
+}
+
+export async function createJWT(
+  header: Partial<JWTHeader> = {},
+  payload: Partial<JWTPayload> = {},
+  address: string,
+  hashingFunction: HashingFucntion
+): Promise<string> {
+  if (!header.typ) header.typ = "JWT";
+    
+  const encodedPayload = (typeof payload === "string") ? payload : encodeSection(payload);
+  const signingInput: string = [encodeSection(header), encodedPayload].join(
+    "."
+  );
+
+  const hash = hashingFunction(signingInput);
+  const bytes = hexToBytes(hash);
+
+  let signature = /* eth_sign MetaMask RPC method that creates the signature (showing the user header and payload that will be signed) */;
+  signature = hexToBytes(signature.slice(0, -2)); // remove byte appended by MetaMask
+  const encodedSignature = bytesToBase64url(signature);
+
+  return [signingInput, encodedSignature].join(".");
+}
+```
+
+### Snap Implementation
+
+Any snap that needs the capability of creating JWTs can do that by calling the JSON-RPC method in the following way:
+
+```typescript
+import {sha256} from 'js-sha256';
+
+const jwt = await wallet.request({
+  method: 'snap_signJwt',
+  params: [
+    {
+      header:  
+        {
+            "alg": "ES256K",
+            "typ": "JWT"
+        },
+      payload: 
+        {
+            "iss": "Company",
+            "sub": "Alice",
+            "role": "employee"
+        },
+      address: '0x12345...',
+      hashingFunction: sha256 // a custom hashing function can be used instead
```

Comment:
This won't work unfortunately. JSON-RPC,  as implied by the name, use JSON. JavaScript functions aren't JSON serialisable.

Reply:
Ah, true. Then the way to go is to pass the hashing function name as a parameter (enum style, e.g., `hashingFunction: 'sha256'`) and use the correct hashing function accordingly:

```typescript
switch (hashingFunction) {
    case 'sha256':
        // sha256 hash
    case 'keccak256':
        // keccak256 hash
    ...
}
```

I think all hashing functions from the library 'ethereum-cryptography' (which is used in `eth-sig-util`) can be available: https://github.com/ethereum/js-ethereum-cryptography#hashes-sha256-keccak-256-ripemd160-blake2b